JavaFX Output Implementation - V2

SonarSonic opened the pull request:
Hey Jannis,

Here is the second version of the JavaFX output implementation, I've created a fresh pull request to resolve the re-basing issues with the master branch and to show the changes more clearly.

Best wishes,
Ollie

SonarSonic committed:
Allow access to paint clean-up in GraphicsUtil

SonarSonic committed:
Deprecate Output.setClip() for removal

SonarSonic committed:
Upgrade batik to a modular version

SonarSonic committed:
Create JavaFX sub module

SonarSonic committed:
Add AWT to JFX inter-op bridge

SonarSonic committed:
Add FXOutput implementation

SonarSonic committed:
Add FXSVGCanvas + FXSVGCanvasSkin + renderer implementations

SonarSonic committed:
Add FXOutput unit testing

SonarSonic committed:
Add JavaFX test viewer application

SonarSonic committed:
Add package-info for jsvg-javafx packages

SonarSonic committed:
Make relevant packages available to jsvg.javafx module

weisJ committed:
Make FXOutput final

weisJ committed:
Add more nullability annotations

weisJ committed:
Expose FXOutput

weisJ committed:
Turn FXSVGRenderer into an interface

weisJ committed:
Fixup expose FXOutput

weisJ committed:
Remove unused function

SonarSonic commented:
@MrKuip Thanks for the suggestion, I think that's something I could look at adding when I'm next working on this, most likely by allowing user defined width + height on the FXSVGCanvas.

weisJ committed:
Move Javafx renderer implementations into own package

weisJ committed:
Add convenience task to run JavaFX test application

weisJ committed:
Make JavaFX test class non-public

weisJ committed:
Wrap newSingleThreadExecutor with try-with-resource

weisJ reviewed (commented):
I have made some changes in my version of the branch: https://github.com/weisJ/jsvg/tree/javafx-final

Looking very good now 🚀

weisJ commented on jsvg-javafx/src/main/java/com/github/weisj/jsvg/ui/jfx/skin/FXSVGCanvasSkin.java:
```diff
@@ -0,0 +1,138 @@
+    private final FXSVGCanvas svgCanvas;
+    private final StackPane innerPane;
+
+    private final AnimationTimer timer;
```
Should animations already work? For me the animation timer stops immediately and no changes in the canvas happen.

SonarSonic commented on gradle.properties:
```diff
@@ -16,18 +16,20 @@ me.champeau.jmh.version                                   = 0.6.6
 javaxAnnotations.version                                  = 1.3.2
 osgiAnnotations.version                                   = 2.0.0
 slf4jApi.version                                          = 2.0.17
+javafx.version                                            = 17.0.15
```
They're slightly out of sync, so this actually only requires Java 11

SonarSonic committed:
Fix JavaFX Test Initialization

SonarSonic committed:
Fix FXSVGCanvas not loading the SVGDocument's animation

SonarSonic committed:
Fix Spotless formatting

weisJ commented on jsvg-javafx/src/test/java/com/github/weisj/jsvg/renderer/jfx/FXOutputTest.java:
```diff
@@ -0,0 +1,182 @@
+import com.github.weisj.jsvg.renderer.output.Output;
+import com.github.weisj.jsvg.view.FloatSize;
+
+public class FXOutputTest {
```
Oh wow, what a footgun! 😅 Directly using a thread is fine. Maybe add a comment about why we don't properly join the thread.

SonarSonic commented:
Hey @weisJ 
I think the implementation was in a good state, I think the main remaining issue was the resizing issue I mentioned in my last message. I'd be very happy for you to continue the work from here, as I don't have much time to work on this at the moment.

weisJ committed:
Hide implementation of FXOutput

weisJ committed:
Run spotless

Copilot committed:
Fix animation invalidations, viewBox/mask issues, and reduce public API

- Fix 1: Only start animation timeline when the document has an animation
  (duration > 0), avoiding unnecessary canvas redraws for static SVGs
- Fix 2: Size the canvas/image to the viewBox dimensions when a viewBox is
  provided, fixing transforms and masks with custom viewBoxes
- Fix 3: Change DEFAULT_USE_SVG_VIEW_BOX to true so the FXSVGCanvas uses
  the document's own viewBox by default
- Fix 4: Move FXSVGRenderer interface and implementations (FXSVGRendererAWT,
  FXSVGRendererJavaFX) to the skin package as package-private, removing them
  from the public API. Also remove unused timeline field from FXSVGCanvasSkin.

Agent-Logs-Url: https://github.com/weisJ/jsvg/sessions/56dc6804-428e-41b0-a0fc-fd383c39ac38

Co-authored-by: weisJ <31143295+weisJ@users.noreply.github.com>

Copilot committed:
Add Javadoc explaining default useSVGViewBox behavior

Agent-Logs-Url: https://github.com/weisJ/jsvg/sessions/56dc6804-428e-41b0-a0fc-fd383c39ac38

Co-authored-by: weisJ <31143295+weisJ@users.noreply.github.com>

weisJ committed:
Make svg be size of full canvas by default

weisJ committed:
Improve demo fx viewer application

weisJ committed:
Format svg-viewer.fxml

weisJ committed:
Add animated svg sample

weisJ committed:
Fix JavaFX tests not running

weisJ commented:
I finally got around to working out the last few blockers. It probably still isn't perfect but at least it is in a state of which I feel comfortable to release a version of. Again thank you so much for working on the initial implementation of this.

SonarSonic commented:
Hey @weisJ 

That's great news, glad you've been able to merge it.
I'll hopefully be able to do some more testing on it in the future!
